// axios-token-handler/build/utils/cookie-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CookieAdapter = void 0;
const length$ = Symbol("document.cookie.length");
class CookieAdapter {
    cookieObject = { [length$]: 0 };
    constructor() {
        if (typeof window !== "undefined") {
            document.cookie
                .split(";")
                .map((c) => c.trim())
                .map((c) => c.split("="))
                .forEach(([key, value]) => {
                if (!key)
                    return;
                this.cookieObject[key] = value || "";
                this.cookieObject[length$]++;
            });
        }
    }
    get length() {
        return this.cookieObject[length$];
    }
    setItem(key, value) {
        if (!key)
            return;
        this.cookieObject[key] = value || "";
        this.cookieObject[length$]++;
        this.save();
    }
    getItem(key) {
        return this.cookieObject[key];
    }
    removeItem(key) {
        if (!key)
            return;
        if (this.cookieObject[key])
            delete this.cookieObject[key];
        this.save();
    }
    clear() {
        this.cookieObject = { [length$]: 0 };
        this.save();
    }
    key(index) {
        throw new Error("Doesn't implemented yet. Try another variant");
    }
    save() {
        if (typeof window !== "undefined") {
            document.cookie = Object.entries(this.cookieObject).reduce((acc, [key, value], idx) => (acc += `${idx === 0 ? "" : ";"}${key}=${value}`), "");
        }
    }
}
exports.CookieAdapter = CookieAdapter;
